### ISSUE TITLE
features that depend upon features (composite features)

### ISSUE THREAD

**Issue by johnnyreilly** [devcontainers/spec#109]
Heya!

I'm a keen user of dapr.  I generally do this in the context of a devcontainer.  I'm interested in how to migrate to doing the same using features.  If I read things right, I need to use the `docker-in-docker` feature.  On top of that I think I'd need to have some kind of `dapr` feature which would contain something like this:

```
RUN apt-get update \
    && apt-get install python3-pip -y \
    # Install Dapr
    && wget -q https://raw.githubusercontent.com/dapr/cli/master/install/install.sh -O - | /bin/bash \
    # Clean up
    && apt-get autoremove -y && apt-get clean -y && rm -rf /var/lib/apt/lists/* 

# Add daprd to the path for the VS Code Dapr extension.
ENV PATH="${PATH}:/home/${USERNAME}/.dapr/bin"
```

I'm guessing that there's some way to have "dependencies" for features?  So the dapr feature would depend upon docker-in-docker.  Am I headed in the right direction here?  Or am I ... not?

It's possible that this is related: https://github.com/devcontainers/cli/issues/163

**Comment by AlexanderLanin** [comment]
I was missing hard-dependencies since I started experimenting with features, but actually that would be a bad idea. The whole point of features is to avoid god like Dockerfiles. We want small reusable pieces. Dependencies between features make it impossible to use a different feature to install python. Or maybe it's already in my base image and I don't even need a feature at all.

My solution for the moment is to detect whether python is installed and in the correct version. If not, my feature exits with an error message telling the user to add python to the Dockerfile or to use \*any\* python feature.

Assuming features are named reasonably, what I would love is to have `installsAfter: "*python*"`

**Comment by joshspicer** [comment]
> Assuming features are named reasonably, what I would love is to have installsAfter: "*python*"

Using a regex is an interesting idea.  It would be assuming a lot, but I bet it would be right the majority of the time (and I suppose _any_ hint is better than random?)

cc/ @chrmarti  If not regexes, could be interesting to explore how else to drop general hints about install order.

**Comment by AlexanderLanin** [comment]
How about fallback to dependencies? Maybe that's some middle ground between hard dependencies and soft dependency order?

What I currently do:
```sh
if ! type python3 >/dev/null 2>&1; then
  echo "ERROR: Use a base image with python3 installed, add it to your Dockerfile or use a python feature like https://github.com/devcontainers/features/tree/main/src/python"
  exit 1
fi
```

How about a silent fallback, somehow supported by cli?
```sh
if ! type python3 >/dev/null 2>&1; then
  echo "INFO: python3 not installed, adding default python installation..."
  import ghcr.io/devcontainers/features/python:1
fi
```

**Comment by gund** [comment]
Hey, I just want to add one more use-case to the "features composition" proposal.
Basically I would like to be able to create an feature aggregate that will simply install other feature, you can think of this as having a "preset" feature with maybe options to set versions and/or options of the features it's going to install.
This is basically what I would want to do to create my own custom dev environment in a single feature that I can take as a base one and optionally add more features on top of it in different projects.

So maybe nesting infinitely many features inside features is not a very nice idea then we could come up with the "preset features" that have a special ability to simply configure which feature does it consists of and what it's default versions/options are, and then on the consuming side a user would be able to configure each sub-feature version/options as needed when using this "preset feature".

What do you think about this idea?

**Comment by joshspicer** [comment]
Hey @danielbraun89 - I've been taking a lot of inspiration from your excellent work on dependencies in `devcontainers-contrib`.  

I'm curious if you have any learnings you'd like to share - I would be very interested to learn what you've learned through the process.

**Comment by joshspicer** [comment]
An implementation based on https://github.com/devcontainers/spec/pull/234 has been merged and shipped with version 0.44.0 of the CLI.  GitHub Codespaces and the Dev Containers extension will pick up the changes in the coming weeks (Dev Containers extension is released alongside the VS Code release cycle).  

If you'd like to take advantage of these changes new changes, i'd suggest iterating on your Features directly with the dev container CLI, and staging the changes for when the functionality is available in the
